src/main/java/hudson/plugins/throttleconcurrents/ThrottleQueueTaskDispatcher.java

```diff
@@ -178,6 +188,32 @@ else if (tjp.getThrottleOption().equals("category")) {
         return null;
     }
 
+    private boolean isBlockedByCategories(List<String> categoryNames) {
+        for (String catName : categoryNames) {
+            if (areProjectsInCategoryBuilding(catName)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private boolean areProjectsInCategoryBuilding(String categoryName) {
+        List<AbstractProject<?, ?>> projectsInCategory = ThrottleJobProperty.getCategoryProjects(categoryName);
+        for (AbstractProject<?, ?> project : projectsInCategory) {
+            if (isProjectBuilding(project)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private boolean isProjectBuilding(AbstractProject<?, ?> project) {
+        if (project.isBuilding() || project.isInQueue()) {
```

Comment:
<code>project.isBuilding()</code> is unreliable for parallel runs; <code>project.isInQueue()</code> causes a full queue lookup with a serious performance impact on cores till 1.610